MINOR: docs(config providers): adds a description of configuration providers

PaulRMellor opened the pull request:
**Documentation**
Adds a section to the docs describing configuration providers and the parameters to use with them, plus a simple example to highlight their use.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

PaulRMellor committed:
docs(config providers): adds a description of configuration providers

Signed-off-by: prmellor <pmellor@redhat.com>

mimaison commented on docs/toc.html:
```diff
@@ -22,61 +22,62 @@
-                <li><a href="#ecosystem">1.4 Ecosystem</a>
-                <li><a href="#upgrade">1.5 Upgrading</a>
-                <li><a href="#docker">1.6 Docker</a>
+                <li><a href="#introduction">1.1 Introduction</a></li>
```
As per the HTML spec, the closing `li` tag is optional: https://www.w3.org/TR/2012/WD-html-markup-20121025/li.html#li-tags

So let's undo these changes.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+    Use configuration providers to load configuration data from external sources. This might include sensitive information, such as passwords, API keys, or other credentials.
+  </p>
+  <p>
+    These configurations can be stored securely and dynamically retrieved as needed. Configuration providers are primarily used with Kafka Connect. For example, you can use providers to supply the credentials for Kafka Connect connectors.
```
`Configuration providers are primarily used with Kafka Connect.`: I don't think this is true. Configuration providers are used by all components. I wonder if we could drop this paragraph altogether.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+  <ul>
+    <li>
+      Use a custom provider by creating a Java class implementing the
+      <code>ConfigProvider</code> interface and packaging it into a JAR file.
```
Could we link to the javadoc for `ConfigProvider`?
To point to the latest javadoc, you can use `{{version}}/javadoc/org/apache/kafka/common/config/provider/ConfigProvider.html`

mimaison replied:
Same below for the built-in providers

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+database.password=${file:/path/to/connector.properties:dbPassword}</code></pre>
+  <p>The configuration provider reads and extracts the values from the properties file.</p>
+  
+    <h3 class="anchor-heading"><a id="java" class="anchor-link"></a><a href="#java">6.6 Java Version</a></h3>
```
We added a new section at 6.6, so we need to bump all section numbers below

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+    Here’s an example that uses a file configuration provider with Kafka Connect to provide authentication credentials to a database for a connector.
+  </p>
+  <p>
+    First, create a <code>connector.properties</code> configuration file with the following credentials:
```
Should we name that file `credentials.properties`, or `connector-credentials.properties` so it's not confused with the actual connector configurations that can also be provided as a file.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,168 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+  
+  <p>Next, reference the properties from the file in the connector configuration.</p>
+  <p><b>Example connector configuration referencing file properties</b></p>
+  <pre><code>database.user=${file:/path/to/connector.properties:dbUsername}
```
In this example, `file` should be `fileProvider` as the first part of the configuration provider magic string is the alias given to the provider.

mimaison reviewed (commented):
Thanks for the PR! I left a few comments and suggestions.

PaulRMellor committed:
docs(review): edits to doc updates from comments by MM

Signed-off-by: prmellor <pmellor@redhat.com>

PaulRMellor commented:
Thansks for the comments @mimaison

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+  <p>You have the following options:</p>
+  <ul>
+    <li>
+      Use a custom provider by creating a Java class implementing the
```
I think we can remove `Java`. You could also use Scala for example to implement a custom ConfigProvider

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+  </h4>
+  
+  <p>
+    The <code>ConfigProvider</code> is a base interface for all configuration providers. Custom implementations of this interface can retrieve configuration data from various sources. You can then package and deploy the implementation as a JAR file and reference the provider's class in your configuration.
```
I think `deploy` is not the right term here. You just need to add the JAR to the classpath.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+  </ul>
+  
+  <p>
+    To use a configuration provider, specify it in your configuration file using the <code>config.providers</code> property. 
```
We can remove `file` in this sentence. You can also build a configuration programmatically.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+  
+  <h4 class="anchor-heading">
+    <a id="common_provider_params" class="anchor-link"></a>
+    <a href="#common_provider_params">Common Parameters</a>
```
What about renaming this to something like "Using Configuration Providers"? as effectively this describe how to use them.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+    To use a configuration provider, specify it in your configuration file using the <code>config.providers</code> property. 
+  </p>
+  <p>
+    Each provider can have its own set of parameters, which are passed in a specific format.
```
If we rename the section below, we could move this paragraph in there.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+config.providers.provider1.class=com.example.Provider1
+config.providers.provider2.class=com.example.Provider2</code></pre>
+  <p>
+    Pass additional parameters to each provider as <code>config.providers.&lt;provider_alias&gt;.&lt;parameter_key&gt;=&lt;parameter_value&gt;</code>:
```
The syntax is not correct. It should be `config.providers.<alias>.param.<name>=<value>`.

Some of the examples below are missing `.param.` too.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+  <p>
+    First, create a <code>connector-credentials.properties</code> configuration file with the following credentials:
+  </p>
+  <p><b>Example configuration file</b></p>
```
I think we can remove this paragraph.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+<pre><code>dbUsername=my-username
+dbPassword=my-password</code></pre>
+  <p><em>NOTE: Ensure the appropriate permissions are set for securely accessing the properties file.
+    For additional security, consider encrypting sensitive information using a secret and implementing decryption in your application at runtime.
```
This tip is not really applicable so I wonder if we drop it. If you're using a configuration provider to provide JAAS credentials for a consumer for example, you can't have some runtime decryption logic. The retrieved values from the provider are directly used.

mimaison commented on docs/ops.html:
```diff
@@ -1247,7 +1247,165 @@ <h4 class="anchor-heading"><a id="prodconfig" class="anchor-link"></a><a href="#
+  <p><b>Example Kafka Connect configuration with an <code>FileConfigProvider</code></b></p>
+  <pre><code>config.providers=fileProvider
+config.providers.fileProvider.class=org.apache.kafka.common.config.provider.FileConfigProvider
+config.providers.fileProvider.config.path=/path/to/connector-properties</code></pre>
```
This line is not correct and should be removed. You don't provide the path when defining the provider, you only provide it when using the provider to retrieve a value.

mimaison reviewed (commented):
Thanks for the updates. I left a few comments on the content itself.

I also wonder if we should move this section at the end of Configuration instead of having it in the middle of the Operations section. WDYT?

PaulRMellor committed:
docs(review): edits to doc from comments by MM -- 02

Signed-off-by: prmellor <pmellor@redhat.com>

PaulRMellor commented:
Thanks for the additional comments @mimaison 
The content now sits in the config section.

mimaison commented on docs/toc.html:
```diff
@@ -50,13 +50,23 @@
+                <li><a href="#tieredstorageconfigs">3.10 Tiered Storage Configs</a>
+                <li><a href="#config_providers">3.11 Configuration Providers</a>
+                    <ul>
+                        <li><a href="#using_providers">Using</a>
```
Should this be `Using Configuration Providers` to match the section title?

PaulRMellor replied:
I think it's fine as it is. I tried a full title with this and referencing and they looked too long in the toc and the concept gets a bit lost. 

![image](https://github.com/user-attachments/assets/dc510ce4-669f-4fdf-a0bf-9cd62fd5f714)

with full names:

![image](https://github.com/user-attachments/assets/4b4153dc-02c4-4714-9c41-4c202cd5fcd5)

mimaison replied:
I'm not a native English speaker and just "Using" looks weird to me. If we don't want the full title because it's too long, would "Usage" work? If you think "Using" is fine, then I'll trust you.

PaulRMellor replied:
Usage is okay but a bit formal and less direct than Using. Using is more action-oriented and fits with referencing (gerund) later. Most technical documents use gerunds for headings. My vote is Using.

mimaison replied:
Ok, thanks for the explanation

mimaison commented on docs/ops.html:
```diff
@@ -1214,9 +1214,9 @@ <h4 class="anchor-heading"><a id="multitenancy-more" class="anchor-link"></a><a
   </p>
 
 
-  <h3 class="anchor-heading"><a id="config" class="anchor-link"></a><a href="#config">6.5 Kafka Configuration</a></h3>
```
Should we undo the changes in this file as the new section is in `configuration.html`? If so let's undo the changes in `toc.html` too.

mimaison commented on docs/configuration.html:
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
+  
+  <h4 class="anchor-heading">
+    <a id="config_provider" class="anchor-link"></a>
+    <a href="#config_provider">ConfigProvider</a>
```
This section only repeats things already mentioned above. Do we really need it?

mimaison replied:
Let's also remove the header in `toc.html`

mimaison commented on docs/configuration.html:
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
+  
+  <h4 class="anchor-heading">
+    <a id="ref_config_provider" class="anchor-link"></a>
+    <a href="#ref_config_provider">Referencing configuration providers</a>
```
I wonder if it would read better if the syntax to use each provider were in their respective sections? WDYT?

PaulRMellor replied:
I think the split make sense. We're doing this somewhere else. Also, people just looking to see how they reference the external properties can come straight here from the TOC.

mimaison replied:
My thinking is that in most cases (configuring clients, brokers, Streams applications, the Connect runtime), it's the same person doing both actions and doing them at the same time. The only outlier scenario is using configuration providers in a connector configuration as the provider first needs to be defined in the runtime and used at a later time when creating a connector.

mimaison commented on docs/configuration.html:
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
+  <p><b>Example <code>DirectoryConfigProvider</code> configuration</b></p>
+  <pre><code>config.providers=dirProvider
+config.providers.dirProvider.class=org.apache.kafka.common.config.provider.DirectoryConfigProvider
+config.providers.dirProvider.param.dir=/path/to/config/directory</code></pre>
```
This is not correct. The path is only provided when using the provider.

mimaison commented on docs/configuration.html:
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
+  
+  <h4 class="anchor-heading">
+    <a id="directory_config_provider" class="anchor-link"></a>
+    <a href="#directory_config_provider">DirectoryConfigProvider</a>
```
This provider has a configurable parameter. I think it would be nice to mention it here:
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/config/provider/DirectoryConfigProvider.java#L48-L50

mimaison commented on docs/configuration.html:
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
+  
+  <h4 class="anchor-heading">
+    <a id="env_var_config_provider" class="anchor-link"></a>
+    <a href="#env_var_config_provider">EnvVarConfigProvider</a>
```
This provider has a configurable parameter. I think it would be nice to mention it here:
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/config/provider/EnvVarConfigProvider.java#L42-L44

mimaison commented on docs/configuration.html:
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
+  
+  <h4 class="anchor-heading">
+    <a id="file_config_provider" class="anchor-link"></a>
+    <a href="#file_config_provider">FileConfigProvider</a>
```
This provider has a configurable parameter. I think it would be nice to mention it here:
https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/config/provider/FileConfigProvider.java#L45-L47

mimaison commented on docs/configuration.html:
```diff
@@ -308,6 +308,157 @@ <h3 class="anchor-heading"><a id="tieredstorageconfigs" class="anchor-link"></a>
+  <p><b>Example connector configuration referencing file properties</b></p>
+  <pre><code>database.user=${fileProvider:/path/to/connector-credentials.properties:dbUsername}
+database.password=${fileProvider:/path/to/connector-credentials.properties:dbPassword}</code></pre>
+  <p>The configuration provider reads and extracts the values from the properties file.</p>  
```
I wonder if we add `At runtime,` at the beginning of the sentence.

mimaison reviewed (commented):
Thanks for the updates. I left a few more suggestions

PaulRMellor committed:
docs(revert): revert ops.html changes

Signed-off-by: prmellor <pmellor@redhat.com>

PaulRMellor committed:
docs(review): edits to doc from comments by MM -- 03

Signed-off-by: prmellor <pmellor@redhat.com>

PaulRMellor commented:
Thanks @mimaison -- we mention the parameters now

PaulRMellor committed:
docs(review): edits to doc from comments by MM -- 04

Signed-off-by: prmellor <pmellor@redhat.com>

PaulRMellor commented:
Thanks @mimaison -- I've now moved the syntax to sit with provider. I renamed the final section "Example: Referencing Files" and because of this renamed "Using" to "Using Configuration Providers" (I didn't like all the _Configuration Providers_ in the headings when we considered this before. I thought it got in the way.)

mimaison reviewed (approved):
LGTM, thanks for the PR
